File: java-solutions/src/leetcode/P75SortColors.java
package leetcode;

import java.util.Arrays;

/**
 * # 75. Sort Colors
 * Given an array `nums` with `n` objects colored red, white, or blue, sort them in-place so that
 * objects of the same color are adjacent, with the colors in the order red, white, and blue.
 * We will use the integers `0`, `1`, and `2` to represent the color red, white, and blue, respectively.
 * You must solve this problem without using the library's sort function.
 */
public class P75SortColors {

    // swap elements at indices i,j in arr
    private void swap(int [] arr, int i, int j) {
        int t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    public void sortColors(int[] nums) {
        // current head of nums
        int h = 0;
        // current tail of nums
        int t = nums.length - 1;
        // index of current element being examined
        int i = 0;

        while (h < t && i <= t) {
            if (nums[i] == 0) {
                swap(nums, h, i);
                if (nums[h] == 0) {
                    h += 1;
                    i += 1;
                }
            } else if (nums[i] == 2) {
                swap(nums, t, i);
                t -= 1;
            } else {
                i += 1;
            }
        }
    }

    public static void main(String[] args) {
        P75SortColors sol = new P75SortColors();
        int [] nums = {2, 0, 2, 1, 1, 0};
        int [] sorted = {0, 0, 1, 1, 2, 2};
        sol.sortColors(nums);

        assert Arrays.equals(nums, sorted);
    }
}

File: rust-solutions/leetcode/src/bin/p75-sort-colors.rs
/// # 75. Sort Colors
/// Given an array `nums` with `n` objects colored red, white, or blue, sort them in-place so that
/// objects of the same color are adjacent, with the colors in the order red, white, and blue.
///
/// We will use the integers `0`, `1`, and `2` to represent the color red, white, and blue, respectively.
///
/// You must solve this problem without using the library's sort function.

struct Solution;

impl Solution {

    pub fn sort_colors(nums: &mut Vec<i32>) {
        // index to current head of nums
        let mut h = 0;
        // index to current tail of nums
        let mut t = nums.len() - 1;
        // current index in nums
        let mut i = 0;

        while h < t && i <= t {
            if nums[i] == 0 {
                nums.swap(h, i);
                if nums[h] == 0 {
                    h += 1;
                    i += 1;
                }
            }
            else if nums[i] == 2 {
                nums.swap(t, i);
                t -= 1;
            } else {
                i += 1;
            }
        }
    }
}

fn main() {}

#[cfg(test)]
mod tests {
    use crate::Solution;

    #[test]
    fn example1() {
        let mut nums = vec![2, 0, 2, 1, 1, 0];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, vec![0,0,1,1,2,2]);
    }

    #[test]
    fn example3() {
        let mut nums = vec![0, 1, 2];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, vec![0,1,2]);
    }

    #[test]
    fn example4() {
        let mut nums = vec![2, 1, 0];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, vec![0,1,2]);
    }

    #[test]
    fn example5() {
        let mut nums = vec![2, 0, 1];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, vec![0,1,2]);
    }
}